File: python-sim/hftext/frame.py
"""HFText Basic v0.1 frame assembly and parsing."""

from __future__ import annotations

from dataclasses import dataclass

from hftext.codec import (
    BITS_PER_BYTE,
    BITS_PER_SYMBOL,
    MAX_PAYLOAD_SYMBOLS,
    decode_symbols_to_text,
    encode_text_to_symbols,
    pack_symbols_to_bytes,
    unpack_symbols_from_bytes,
)
from hftext.crc16 import crc16_ccitt_false

SYNC_VALUE = 0x2DD4
SYNC_BYTES = SYNC_VALUE.to_bytes(2, "big")
PREAMBLE_BITS = [1, 0] * 32
CRC_BYTES = 2
HEADER_BYTES = 3


@dataclass(frozen=True)
class FrameResult:
    frame_detected: bool
    crc_ok: bool
    payload_valid: bool
    text: str
    length: int = 0
    payload_symbols: list[int] | None = None
    error: str | None = None
    sync_index: int | None = None


def bytes_to_bits(data: bytes) -> list[int]:
    """Return bytes as MSB-first bits."""
    bits = []
    for byte in data:
        for shift in range(BITS_PER_BYTE - 1, -1, -1):
            bits.append((byte >> shift) & 1)
    return bits


SYNC_BITS = bytes_to_bits(SYNC_BYTES)


def bits_to_bytes(bits: list[int]) -> bytes:
    """Return MSB-first bits as bytes."""
    if len(bits) % BITS_PER_BYTE:
        raise ValueError("bit count must be a multiple of 8")

    output = bytearray()
    for offset in range(0, len(bits), BITS_PER_BYTE):
        byte = 0
        for bit in bits[offset : offset + BITS_PER_BYTE]:
            if bit not in (0, 1):
                raise ValueError(f"invalid bit: {bit}")
            byte = (byte << 1) | bit
        output.append(byte)
    return bytes(output)


def payload_byte_count(length: int) -> int:
    """Return the number of packed bytes needed for length symbols."""
    if length < 0 or length > MAX_PAYLOAD_SYMBOLS:
        raise ValueError(f"length must be between 0 and {MAX_PAYLOAD_SYMBOLS}")
    return (length * BITS_PER_SYMBOL + BITS_PER_BYTE - 1) // BITS_PER_BYTE


def build_frame_bytes(payload_text: str) -> bytes:
    """Build an HFText frame as bytes."""
    payload_symbols = encode_text_to_symbols(payload_text)
    length = len(payload_symbols)
    payload = pack_symbols_to_bytes(payload_symbols)
    crc = crc16_ccitt_false(payload).to_bytes(CRC_BYTES, "big")

    return SYNC_BYTES + bytes([length]) + payload + crc


def build_frame(payload_text: str) -> list[int]:
    """Build an HFText frame as MSB-first bits."""
    return bytes_to_bits(build_frame_bytes(payload_text))


def build_transmission(payload_text: str, preamble_bits: list[int] | None = None) -> list[int]:
    """Build TX bits with a preamble followed by the HFText frame."""
    preamble = PREAMBLE_BITS if preamble_bits is None else preamble_bits
    for bit in preamble:
        if bit not in (0, 1):
            raise ValueError(f"invalid preamble bit: {bit}")
    return list(preamble) + build_frame(payload_text)


def parse_frame_bytes(frame: bytes) -> FrameResult:
    """Parse an exact HFText frame from bytes."""
    if len(frame) < HEADER_BYTES + CRC_BYTES:
        return FrameResult(False, False, False, "", error="frame too short")

    if frame[:2] != SYNC_BYTES:
        return FrameResult(False, False, False, "", error="sync not found")

    length = frame[2]
    if length & 0x80:
        return FrameResult(True, False, False, "", length=length, error="length bit 7 set")
    if length > MAX_PAYLOAD_SYMBOLS:
        return FrameResult(True, False, False, "", length=length, error="length too large")

    packed_size = payload_byte_count(length)
    expected_size = HEADER_BYTES + packed_size + CRC_BYTES
    if len(frame) != expected_size:
        return FrameResult(
            True,
            False,
            False,
            "",
            length=length,
            error=f"frame size mismatch: expected {expected_size}, got {len(frame)}",
        )

    payload = frame[HEADER_BYTES : HEADER_BYTES + packed_size]
    received_crc = int.from_bytes(frame[HEADER_BYTES + packed_size :], "big")
    calculated_crc = crc16_ccitt_false(payload)
    crc_ok = received_crc == calculated_crc
    if not crc_ok:
        return FrameResult(True, False, False, "", length=length, error="crc mismatch")

    payload_symbols = unpack_symbols_from_bytes(payload, length)
    try:
        text = decode_symbols_to_text(payload_symbols)
    except ValueError as exc:
        return FrameResult(
            True,
            True,
            False,
            "",
            length=length,
            payload_symbols=payload_symbols,
            error=str(exc),
        )

    return FrameResult(True, True, True, text, length=length, payload_symbols=payload_symbols)


def parse_frame(bits: list[int]) -> FrameResult:
    """Parse an exact HFText frame from MSB-first bits."""
    return parse_frame_bytes(bits_to_bytes(bits))


def find_sync(bits: list[int]) -> int | None:
    """Return the bit index of SYNC in a received bit stream."""
    if len(bits) < len(SYNC_BITS):
        return None
    for bit in bits:
        if bit not in (0, 1):
            raise ValueError(f"invalid bit: {bit}")

    last_start = len(bits) - len(SYNC_BITS)
    for index in range(last_start + 1):
        if bits[index : index + len(SYNC_BITS)] == SYNC_BITS:
            return index
    return None


def _bits_to_int(bits: list[int]) -> int:
    value = 0
    for bit in bits:
        if bit not in (0, 1):
            raise ValueError(f"invalid bit: {bit}")
        value = (value << 1) | bit
    return value


def _parse_frame_at_sync(bits: list[int], sync_index: int) -> FrameResult:
    length_start = sync_index + len(SYNC_BITS)
    length_end = length_start + BITS_PER_BYTE
    if len(bits) < length_end:
        return FrameResult(True, False, False, "", error="stream ended before length", sync_index=sync_index)

    length = _bits_to_int(bits[length_start:length_end])
    if length & 0x80:
        return FrameResult(
            True,
            False,
            False,
            "",
            length=length,
            error="length bit 7 set",
            sync_index=sync_index,
        )
    if length > MAX_PAYLOAD_SYMBOLS:
        return FrameResult(
            True,
            False,
            False,
            "",
            length=length,
            error="length too large",
            sync_index=sync_index,
        )

    frame_bit_count = (HEADER_BYTES + payload_byte_count(length) + CRC_BYTES) * BITS_PER_BYTE
    frame_end = sync_index + frame_bit_count
    if len(bits) < frame_end:
        return FrameResult(
            True,
            False,
            False,
            "",
            length=length,
            error="stream ended before complete frame",
            sync_index=sync_index,
        )

    result = parse_frame(bits[sync_index:frame_end])
    return FrameResult(
        result.frame_detected,
        result.crc_ok,
        result.payload_valid,
        result.text,
        length=result.length,
        payload_symbols=result.payload_symbols,
        error=result.error,
        sync_index=sync_index,
    )


def parse_frame_from_stream(bits: list[int]) -> FrameResult:
    """Find SYNC candidates and return the first valid HFText frame."""
    if len(bits) < len(SYNC_BITS):
        return FrameResult(False, False, False, "", error="sync not found")
    for bit in bits:
        if bit not in (0, 1):
            raise ValueError(f"invalid bit: {bit}")

    first_candidate: FrameResult | None = None
    last_start = len(bits) - len(SYNC_BITS)
    for sync_index in range(last_start + 1):
        if bits[sync_index : sync_index + len(SYNC_BITS)] != SYNC_BITS:
            continue

        result = _parse_frame_at_sync(bits, sync_index)
        if first_candidate is None:
            first_candidate = result
        if result.crc_ok and result.payload_valid:
            return result

    if first_candidate is not None:
        return first_candidate

    return FrameResult(False, False, False, "", error="sync not found")

File: core/src/frame.cpp
#include "hftext_frame.h"

#include "hftext_crc16.h"
#include "hftext_encoder.h"

#include <algorithm>
#include <stdexcept>
#include <string>
#include <utility>

namespace hftext {
namespace {

void validateBit(std::uint8_t bit) {
    if (bit != 0 && bit != 1) {
        throw std::invalid_argument("invalid bit");
    }
}

int bitsToInt(const std::vector<std::uint8_t>& bits, std::size_t start, std::size_t count) {
    int value = 0;
    for (std::size_t index = start; index < start + count; ++index) {
        validateBit(bits[index]);
        value = (value << 1) | bits[index];
    }
    return value;
}

std::uint16_t readBigEndianU16(const std::vector<std::uint8_t>& bytes, std::size_t start) {
    return static_cast<std::uint16_t>((static_cast<std::uint16_t>(bytes[start]) << 8) | bytes[start + 1]);
}

void appendBigEndianU16(std::vector<std::uint8_t>& bytes, std::uint16_t value) {
    bytes.push_back(static_cast<std::uint8_t>((value >> 8) & 0xFF));
    bytes.push_back(static_cast<std::uint8_t>(value & 0xFF));
}

DecodeResult makeResult(
    bool frameDetected,
    bool crcOk,
    bool payloadValid,
    std::string text,
    int length = 0,
    std::vector<std::uint8_t> payloadSymbols = {},
    std::string error = {},
    int syncIndex = -1
) {
    DecodeResult result;
    result.frameDetected = frameDetected;
    result.crcOk = crcOk;
    result.payloadValid = payloadValid;
    result.text = std::move(text);
    result.length = length;
    result.payloadSymbols = std::move(payloadSymbols);
    result.error = std::move(error);
    result.syncIndex = syncIndex;
    return result;
}

}  // namespace

std::vector<std::uint8_t> syncBytes() {
    return {0x2D, 0xD4};
}

std::vector<std::uint8_t> syncBits() {
    return bytesToBits(syncBytes());
}

std::vector<std::uint8_t> startSyncBits() {
    auto bits = syncBits();
    const auto repeated = syncBits();
    bits.insert(bits.end(), repeated.begin(), repeated.end());
    return bits;
}

std::vector<std::uint8_t> physicalLengthBits(int payloadLength) {
    if (payloadLength < 0 || payloadLength > kMaxPayloadSymbols) {
        throw std::invalid_argument("payload length must be between 0 and 127");
    }

    std::vector<std::uint8_t> bits;
    bits.reserve(kBitsPerByte * kPhysicalLengthRepeat);
    const auto byte = static_cast<std::uint8_t>(payloadLength & 0x7F);
    for (int repeat = 0; repeat < kPhysicalLengthRepeat; ++repeat) {
        for (int shift = kBitsPerByte - 1; shift >= 0; --shift) {
            bits.push_back(static_cast<std::uint8_t>((byte >> shift) & 1U));
        }
    }
    return bits;
}

int decodePhysicalLengthBits(const std::vector<std::uint8_t>& bits, std::size_t start) {
    const auto requiredBits = static_cast<std::size_t>(kBitsPerByte * kPhysicalLengthRepeat);
    if (start + requiredBits > bits.size()) {
        return -1;
    }

    int value = 0;
    for (int bitIndex = 0; bitIndex < kBitsPerByte; ++bitIndex) {
        int ones = 0;
        for (int repeat = 0; repeat < kPhysicalLengthRepeat; ++repeat) {
            const auto bit = bits[start + static_cast<std::size_t>(repeat * kBitsPerByte + bitIndex)];
            validateBit(bit);
            ones += bit;
        }
        value = (value << 1) | (ones >= 2 ? 1 : 0);
    }

    if ((value & 0x80) != 0 || value > kMaxPayloadSymbols) {
        return -1;
    }
    return value;
}

std::vector<std::uint8_t> defaultPreambleBits() {
    std::vector<std::uint8_t> bits;
    bits.reserve(kDefaultPreambleBits);
    for (int index = 0; index < kDefaultPreambleBits; ++index) {
        bits.push_back(static_cast<std::uint8_t>(index % 2 == 0 ? 1 : 0));
    }
    return bits;
}

std::vector<std::uint8_t> bytesToBits(const std::vector<std::uint8_t>& data) {
    std::vector<std::uint8_t> bits;
    bits.reserve(data.size() * kBitsPerByte);
    for (std::uint8_t byte : data) {
        for (int shift = kBitsPerByte - 1; shift >= 0; --shift) {
            bits.push_back(static_cast<std::uint8_t>((byte >> shift) & 1U));
        }
    }
    return bits;
}

std::vector<std::uint8_t> bitsToBytes(const std::vector<std::uint8_t>& bits) {
    if (bits.size() % kBitsPerByte != 0) {
        throw std::invalid_argument("bit count must be a multiple of 8");
    }

    std::vector<std::uint8_t> bytes;
    bytes.reserve(bits.size() / kBitsPerByte);
    for (std::size_t offset = 0; offset < bits.size(); offset += kBitsPerByte) {
        std::uint8_t byte = 0;
        for (std::size_t index = offset; index < offset + kBitsPerByte; ++index) {
            validateBit(bits[index]);
            byte = static_cast<std::uint8_t>((byte << 1) | bits[index]);
        }
        bytes.push_back(byte);
    }
    return bytes;
}

int payloadByteCount(int length) {
    if (length < 0 || length > kMaxPayloadSymbols) {
        throw std::invalid_argument("length must be between 0 and 127");
    }
    return (length * kBitsPerSymbol + kBitsPerByte - 1) / kBitsPerByte;
}

std::vector<std::uint8_t> packSymbolsToBytes(const std::vector<std::uint8_t>& symbols) {
    std::uint32_t bitBuffer = 0;
    int bitCount = 0;
    std::vector<std::uint8_t> bytes;
    bytes.reserve(payloadByteCount(static_cast<int>(symbols.size())));

    for (std::uint8_t symbol : symbols) {
        if (symbol >= (1U << kBitsPerSymbol)) {
            throw std::invalid_argument("invalid 6-bit symbol");
        }

        bitBuffer = (bitBuffer << kBitsPerSymbol) | symbol;
        bitCount += kBitsPerSymbol;

        while (bitCount >= kBitsPerByte) {
            bitCount -= kBitsPerByte;
            bytes.push_back(static_cast<std::uint8_t>((bitBuffer >> bitCount) & 0xFF));
            bitBuffer &= (1U << bitCount) - 1U;
        }
    }

    if (bitCount != 0) {
        bytes.push_back(static_cast<std::uint8_t>((bitBuffer << (kBitsPerByte - bitCount)) & 0xFF));
    }

    return bytes;
}

std::vector<std::uint8_t> unpackSymbolsFromBytes(const std::vector<std::uint8_t>& data, int symbolCount) {
    if (symbolCount < 0) {
        throw std::invalid_argument("symbol_count must be non-negative");
    }

    const int availableBits = static_cast<int>(data.size()) * kBitsPerByte;
    const int requiredBits = symbolCount * kBitsPerSymbol;
    if (requiredBits > availableBits) {
        throw std::invalid_argument("not enough data for requested symbols");
    }

    std::uint32_t bitBuffer = 0;
    int bitCount = 0;
    std::vector<std::uint8_t> symbols;
    symbols.reserve(static_cast<std::size_t>(symbolCount));

    for (std::uint8_t byte : data) {
        bitBuffer = (bitBuffer << kBitsPerByte) | byte;
        bitCount += kBitsPerByte;

        while (bitCount >= kBitsPerSymbol && static_cast<int>(symbols.size()) < symbolCount) {
            bitCount -= kBitsPerSymbol;
            symbols.push_back(static_cast<std::uint8_t>((bitBuffer >> bitCount) & 0x3F));
            bitBuffer &= (1U << bitCount) - 1U;
        }
    }

    return symbols;
}

std::vector<std::uint8_t> buildFrameBytes(const std::string& payloadText) {
    const auto payloadSymbols = encodeTextToSymbols(payloadText);
    const auto payload = packSymbolsToBytes(payloadSymbols);
    const std::uint16_t crc = crc16CcittFalse(payload);

    std::vector<std::uint8_t> frame = syncBytes();
    frame.push_back(static_cast<std::uint8_t>(payloadSymbols.size()));
    frame.insert(frame.end(), payload.begin(), payload.end());
    appendBigEndianU16(frame, crc);
    return frame;
}

std::vector<std::uint8_t> buildFrame(const std::string& payloadText) {
    return bytesToBits(buildFrameBytes(payloadText));
}

std::vector<std::uint8_t> buildTransmission(const std::string& payloadText) {
    return buildTransmission(payloadText, defaultPreambleBits());
}

std::vector<std::uint8_t> buildTransmission(
    const std::string& payloadText,
    const std::vector<std::uint8_t>& preambleBits
) {
    for (std::uint8_t bit : preambleBits) {
        validateBit(bit);
    }

    auto transmission = preambleBits;
    const auto frame = buildFrame(payloadText);
    transmission.insert(transmission.end(), frame.begin(), frame.end());
    return transmission;
}

DecodeResult parseFrameBytes(const std::vector<std::uint8_t>& frame) {
    if (frame.size() < kHeaderBytes + kCrcBytes) {
        return makeResult(false, false, false, "", 0, {}, "frame too short");
    }

    const auto expectedSync = syncBytes();
    if (!std::equal(expectedSync.begin(), expectedSync.end(), frame.begin())) {
        return makeResult(false, false, false, "", 0, {}, "sync not found");
    }

    const int length = frame[2];
    if ((length & 0x80) != 0) {
        return makeResult(true, false, false, "", length, {}, "length bit 7 set");
    }
    if (length > kMaxPayloadSymbols) {
        return makeResult(true, false, false, "", length, {}, "length too large");
    }

    const int packedSize = payloadByteCount(length);
    const std::size_t expectedSize = static_cast<std::size_t>(kHeaderBytes + packedSize + kCrcBytes);
    if (frame.size() != expectedSize) {
        return makeResult(
            true,
            false,
            false,
            "",
            length,
            {},
            "frame size mismatch: expected " + std::to_string(expectedSize) + ", got " + std::to_string(frame.size())
        );
    }

    const auto payloadStart = frame.begin() + kHeaderBytes;
    const auto payloadEnd = payloadStart + packedSize;
    const std::vector<std::uint8_t> payload(payloadStart, payloadEnd);
    const std::uint16_t receivedCrc = readBigEndianU16(frame, kHeaderBytes + packedSize);
    const std::uint16_t calculatedCrc = crc16CcittFalse(payload);
    if (receivedCrc != calculatedCrc) {
        return makeResult(true, false, false, "", length, {}, "crc mismatch");
    }

    const auto payloadSymbols = unpackSymbolsFromBytes(payload, length);
    try {
        const auto text = decodeSymbolsToText(payloadSymbols);
        return makeResult(true, true, true, text, length, payloadSymbols);
    } catch (const std::invalid_argument& exc) {
        return makeResult(true, true, false, "", length, payloadSymbols, exc.what());
    }
}

DecodeResult parseFrame(const std::vector<std::uint8_t>& bits) {
    return parseFrameBytes(bitsToBytes(bits));
}

DecodeResult parseFrameAtSync(const std::vector<std::uint8_t>& bits, int syncIndex) {
    const std::size_t lengthStart = static_cast<std::size_t>(syncIndex) + syncBits().size();
    const std::size_t lengthEnd = lengthStart + kBitsPerByte;
    if (bits.size() < lengthEnd) {
        return makeResult(true, false, false, "", 0, {}, "stream ended before length", syncIndex);
    }

    const int length = bitsToInt(bits, lengthStart, kBitsPerByte);
    if ((length & 0x80) != 0) {
        return makeResult(true, false, false, "", length, {}, "length bit 7 set", syncIndex);
    }
    if (length > kMaxPayloadSymbols) {
        return makeResult(true, false, false, "", length, {}, "length too large", syncIndex);
    }

    const std::size_t frameBitCount = static_cast<std::size_t>(kHeaderBytes + payloadByteCount(length) + kCrcBytes)
        * kBitsPerByte;
    const std::size_t frameEnd = static_cast<std::size_t>(syncIndex) + frameBitCount;
    if (bits.size() < frameEnd) {
        return makeResult(true, false, false, "", length, {}, "stream ended before complete frame", syncIndex);
    }

    const std::vector<std::uint8_t> frameBits(
        bits.begin() + syncIndex,
        bits.begin() + static_cast<std::ptrdiff_t>(frameEnd)
    );
    auto result = parseFrame(frameBits);
    result.syncIndex = syncIndex;
    return result;
}

int findSync(const std::vector<std::uint8_t>& bits) {
    const auto sync = syncBits();
    if (bits.size() < sync.size()) {
        return -1;
    }

    for (std::uint8_t bit : bits) {
        validateBit(bit);
    }

    const std::size_t lastStart = bits.size() - sync.size();
    for (std::size_t index = 0; index <= lastStart; ++index) {
        if (std::equal(sync.begin(), sync.end(), bits.begin() + static_cast<std::ptrdiff_t>(index))) {
            return static_cast<int>(index);
        }
    }
    return -1;
}

DecodeResult parseFrameFromStream(const std::vector<std::uint8_t>& bits) {
    const auto sync = syncBits();
    if (bits.size() < sync.size()) {
        return makeResult(false, false, false, "", 0, {}, "sync not found");
    }

    for (std::uint8_t bit : bits) {
        validateBit(bit);
    }

    DecodeResult firstCandidate;
    bool hasCandidate = false;
    const std::size_t lastStart = bits.size() - sync.size();
    for (std::size_t index = 0; index <= lastStart; ++index) {
        if (!std::equal(sync.begin(), sync.end(), bits.begin() + static_cast<std::ptrdiff_t>(index))) {
            continue;
        }

        auto result = parseFrameAtSync(bits, static_cast<int>(index));
        if (!hasCandidate) {
            firstCandidate = result;
            hasCandidate = true;
        }
        if (result.crcOk && result.payloadValid) {
            return result;
        }
    }

    if (hasCandidate) {
        return firstCandidate;
    }

    return makeResult(false, false, false, "", 0, {}, "sync not found");
}

}  // namespace hftext
